ipn/ipnlocal/cert.go

```diff
@@ -122,10 +125,80 @@ func shouldStartDomainRenewal(dir, domain string, future time.Time) bool {
 		return false
 	}
 	lastRenewCheck[domain] = now
-	_, ok := getCertPEMCached(dir, domain, future)
+	_, ok := b.getCertPEMCached(dir, domain, future)
 	return !ok
 }
 
+// certStore provides a way to perist and retrieve TLS certificates.
+// As of 2023-02-01, we use store certs in directories on disk everywhere
+// except on Kubernetes, where we use the state store.
+type certStore interface {
+	Read(domain string, now time.Time) (*TLSCertKeyPair, error)
+	WriteCert(domain string, cert []byte) error
+	WriteKey(domain string, key []byte) error
+}
+
+func (b *LocalBackend) getCertStore(dir string) certStore {
+	if hostinfo.GetEnvType() == hostinfo.Kubernetes {
+		if _, ok := b.store.(*kubestore.Store); ok && dir == "/tmp" {
+			return certStateStore{b.store}
+		}
+	}
+	return certFileStore(dir)
+}
+
+type certFileStore string // dir
+
+func (f certFileStore) Read(domain string, now time.Time) (*TLSCertKeyPair, error) {
+	certPEM, err := os.ReadFile(keyFile(string(f), domain))
+	if err != nil {
+		return nil, err
+	}
+	keyPEM, err := os.ReadFile(certFile(string(f), domain))
+	if err != nil {
+		return nil, err
+	}
+	if !validCertPEM(domain, keyPEM, certPEM, now) {
+		return nil, errors.New("invalid cert")
+	}
+	return &TLSCertKeyPair{CertPEM: certPEM, KeyPEM: keyPEM, Cached: true}, nil
+}
+
+func (f certFileStore) WriteCert(domain string, cert []byte) error {
+	return os.WriteFile(keyFile(string(f), domain), cert, 0644)
+}
+
+func (f certFileStore) WriteKey(domain string, key []byte) error {
+	return os.WriteFile(keyFile(string(f), domain), key, 0600)
+}
+
+type certStateStore struct {
```

Comment:
// certStateStore implements certStore by storing the cert & key files in an ipn.StateStore.

Reply:
done